### PR TITLE
Support pydantic plugin in 2.xx version

### PR DESCRIPTION
**Pull request by Future-Outlier**
## Tracking Issues
https://flyte-org.slack.com/archives/CP2HDHKE1/p1708960554466059

## Why are the changes needed?
When pydantic version >= 2, this current plugin will not work.

## What changes were proposed in this pull request?
- use `as pydantic` to support 2 cases , `pydantic version < 2` and `pydantic version >= 2`
- remove `lazy_load` to support 2 cases , `pydantic version < 2` and `pydantic version >= 2`
## How was this patch tested?
We test it in 2 scenarios, `pydantic version < 2` and `pydantic version >= 2`.
We need to use
- `from pydantic import BaseModel` when `pydantic  version < 2`
- `from pydantic.v1 import BaseModel` when `pydantic  version >= 2`

```python
from flytekit import task
from pydantic.v1 import BaseModel

class TrainConfig(BaseModel):
    lr: float = 1e-3
    batch_size: int = 32

@task
# def train() -> TrainConfig:
def train() -> TrainConfig:
    return TrainConfig(lr=1.1, batch_size=64)

if __name__ == "__main__":
    print(train(cfg=TrainConfig(lr=1.1, batch_size=64)))
```
```Dockerfile
FROM python:3.9-slim-buster
USER root
WORKDIR /root
ENV PYTHONPATH /root
RUN apt-get update && apt-get install build-essential -y
RUN apt-get install git -y

RUN pip install -U git+https://github.com/flyteorg/flytekit.git@e85e6ec5dba79dc09791abd1f1eac1ce115f0ef6#subdirectory=plugins/flytekit-pydantic

RUN pip install -U git+https://github.com/flyteorg/flytekit.git@e85e6ec5dba79dc09791abd1f1eac1ce115f0ef6
```

```bash
pyflyte run --image futureoutlier/flytekit:pydantic-v2 --remote pydantic_example.py train
```

#### when pydantic version < 2
```python
from flytekit import task
from pydantic import BaseModel
```
image

#### when pydantic version >= 2
```python
from flytekit import task
from pydantic.v1 import BaseModel
```
image

#### remote execution
image

#### warning message
image

## Check all the applicable boxes 

- [ ] I updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

### CODE REVIEW

File: .github/workflows/pythonbuild.yml [hunk 1]
```diff
@@ -242,12 +244,12 @@ jobs:
           - flytekit-pandera
           - flytekit-papermill
           - flytekit-polars
+          - flytekit-pydantic
```
Comment (pingsutw) [inline .github/workflows/pythonbuild.yml:247]: thank you, I don't know why we missed this.

File: plugins/flytekit-pydantic/setup.py [hunk 2]
```diff
@@ -4,7 +4,7 @@
 
 microlib_name = f"flytekitplugins-{PLUGIN_NAME}"
 
-plugin_requires = ["flytekit>=1.7.0b0,<2.0.0", "pydantic"]
+plugin_requires = ["flytekit>=1.7.0,<2.0.0", "pydantic<2.0.0"]
```
Comment (pingsutw) [inline plugins/flytekit-pydantic/setup.py:7]: why do we need to change here

Comment (Future-Outlier) [reply plugins/flytekit-pydantic/setup.py]: Thank you for questioning this, just come out a better solution!

File: plugins/flytekit-pydantic/flytekitplugins/pydantic/basemodel_transformer.py [hunk 3]
```diff
@@ -5,13 +5,18 @@
...
 
-from flytekit import FlyteContext, lazy_module
+from flytekit import FlyteContext
 from flytekit.core import type_engine
 from flytekit.models import literals, types
 
 from . import deserialization, serialization
 
-pydantic = lazy_module("pydantic")
+# pydantic plugin is not suitable for lazy module.
```
Comment (pingsutw) [inline plugins/flytekit-pydantic/flytekitplugins/pydantic/basemodel_transformer.py:14]:
```suggestion
```

File: plugins/flytekit-pydantic/flytekitplugins/pydantic/basemodel_transformer.py [hunk 4]
```diff
@@ -5,13 +5,18 @@
...
+try:
+    # it will cause error if it is lesser than 2.0 version
```
Comment (pingsutw) [inline plugins/flytekit-pydantic/flytekitplugins/pydantic/basemodel_transformer.py:16]:
```suggestion
    # TODO: Use pydantic v2 to serialize/deserialize data
    # <Link to the issue>
```

File: plugins/flytekit-pydantic/tests/test_type_transformer.py [hunk 5]
```diff
@@ -8,14 +8,19 @@
...
-from pydantic import BaseModel, Extra
 
 import flytekit
 from flytekit.core import context_manager
 from flytekit.core.type_engine import TypeEngine
 from flytekit.types import directory
 from flytekit.types.file import file
 
+try:
+    # it will cause error if it is lesser than 2.0 version
```
Comment (pingsutw) [inline plugins/flytekit-pydantic/tests/test_type_transformer.py:19]: ditto

File: plugins/flytekit-pydantic/flytekitplugins/pydantic/deserialization.py [hunk 6]
```diff
@@ -3,15 +3,19 @@
...
 from flytekitplugins.pydantic import commons, serialization
 
-from flytekit import lazy_module
 from flytekit.core import context_manager, type_engine
 from flytekit.models import literals
 from flytekit.types import directory, file
 
-pydantic = lazy_module("pydantic")
+try:
+    # it will cause error if it is lesser than 2.0 version
```
Comment (pingsutw) [inline plugins/flytekit-pydantic/flytekitplugins/pydantic/deserialization.py:11]: ditto

File: plugins/flytekit-pydantic/flytekitplugins/pydantic/serialization.py [hunk 7]
```diff
@@ -14,13 +14,16 @@
...
 
-from flytekit import lazy_module
 from flytekit.core import context_manager, type_engine
 from flytekit.models import literals
 
 from . import commons
 
-pydantic = lazy_module("pydantic")
+try:
+    # it will cause error if it is lesser than 2.0 version
```
Comment (pingsutw) [inline plugins/flytekit-pydantic/flytekitplugins/pydantic/serialization.py:23]: ditto

**Comment by Future-Outlier** [comment]
### will not show a warning message when import flytekit
image

### will show warning message when using pydantic plugin
image

**Review by pingsutw — APPROVED** [review]
LGTM, thanks

**Pull request merged**